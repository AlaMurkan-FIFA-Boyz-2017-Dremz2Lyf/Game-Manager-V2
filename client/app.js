// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import promise from 'redux-promise';

// Components
import Home from './components/home';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(promise)
));

ReactDOM.render(
  <Provider store={store}>
    <Home/>
  </Provider>,
  document.getElementById('app')
);
