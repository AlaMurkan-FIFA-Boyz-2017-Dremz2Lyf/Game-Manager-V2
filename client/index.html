<html>
<head>
  <title>Game Manager</title>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans|Oswald|PT+Sans">
</head>
<body>
  <div id="app"></div>
</body>
</html>
