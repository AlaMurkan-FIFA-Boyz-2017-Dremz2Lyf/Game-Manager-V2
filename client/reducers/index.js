import { combineReducers } from 'redux';
import data from './data_reducer';
import isError from './isError';
import isLoading from './isLoading';

const rootReducer = combineReducers({
  data,
  isError,
  isLoading
});

export default rootReducer;
