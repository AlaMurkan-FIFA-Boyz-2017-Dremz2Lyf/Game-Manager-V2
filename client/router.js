import { Router, browserHistory } from 'react-router';
import React from 'react';

import Home from './components/home';
import App from './components/app';


const routeConfigs = {
  component: App,
  path: '/',
  indexRoute: { component: Home },
  childRoutes: [
    {
      path: 'play/:id',
      getComponent(location, cb) {
        System.import('./components/play_tournament').then(module => cb(null, module.default)).catch(err => cb(err, null));
      }
    }
  ]
};

const Routes = () => (
  <Router history={browserHistory} routes={routeConfigs}/>
);

export { Routes as default };
