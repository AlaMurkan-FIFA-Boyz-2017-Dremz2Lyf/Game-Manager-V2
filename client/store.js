
// Make our store exportable for using in tests/and production

import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(thunk)
));
