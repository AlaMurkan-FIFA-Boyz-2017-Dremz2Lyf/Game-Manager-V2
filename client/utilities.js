/**
  * normalize
  * @param {array} list - an array of objects to be normalized into an object
*/
export const normalize = (list) => {

  return list.reduce((normalized, item) => {
    normalized[item.id] = item;
    return normalized;
  }, {});

};

/**
  * applyAll - apply all uses a bit of tricky programming here to return a new state adding to only the specific substate we need.
  * @param {object} state - the current state
  * @param {object} action - the action coming through, it's payload must have a stateKey and the data corresponding to that statekey.
*/
export const applyAll = (state, action) => {
  const { stateKey, data } = action.payload;

  return {...state, [stateKey]: data};
};

/**
  * percentPlayed - basic math here.
  * @param {number} gamesPlayed - pretty self explanatory
  * @param {number} totalGames - pretty self explanatory
*/
export const percentPlayed = (gamesPlayed, totalGames) => {
  return Math.floor((gamesPlayed / totalGames) * 100);
};

/**
  * getValidationState - will give us the correct bootstrap class for UI responsiveness
  * @param {bool} valid - passed from redux-form, lets us know if the form is valid
  * @param {bool} pristine - from redux-form, lets us know if the form has been touched & edited
*/
export const getValidationState = (valid, pristine) => pristine ? null : (valid ? 'success' : 'error' );

/**
  * required - another validation check for redux-form
  * @param {string} value - from redux-form, returns an alert if the user has not entered required info. Also deals with the number 0 as that is falsey value. 
*/
export const required = value => value === 0 || value ? undefined : 'Required';

/**
  * possValidation - validation specific to the possession field
  * @param {string} value - the input from the form
  * @param {string} allValues - all the inputs from the form
*/
export const possValidation = (value, allValues) => {
  const { p1Poss, p2Poss } = allValues;

  if (!p1Poss && !p2Poss) {
    return 'Possession is required.';
  } else if (Number(p1Poss) + Number(p2Poss) !== 100) {
    return 'Possession should add up to 100%!';
  }
  return undefined;
};

/**
  * helpMessage - renders the correct help message based on the field
  * @param {string} field
*/
export const helpMessage = (field) => {
  let message;

  if (field === 'Score') {
    message = `${field} is required! I mean.. come on!`;
  } else {
    message = `${field} will default to 0 if not given.`;
  }
  return message;
};

/**
  * buildTable
  * @param {object} games - a normalized games object to build the table from
  * @param {object} players - another normalized object to reference the player names.
*/
export const buildTable = (games, players) => {

  let trackedStats = {
    wins: 0, losses: 0, draws: 0, goalsFor: 0, goalsAgainst: 0, shots: 0, onGoal: 0, reds: 0, yellows: 0, poss: null, points: 0, goalDiff: 0
  };

  return Object.values(games).reduce((table, game) => {

    let { p1, p2 } = game;

    table[p1] = table[p1] ? table[p1] : { ...trackedStats, id: p1, username: players[p1].username};
    table[p2] = table[p2] ? table[p2] : { ...trackedStats, id: p2, username: players[p2].username};

    return applyGame(table, game);
  }, {});
};


// I extracted this out of the reduce in build table because of all the action that needs to happen.
// There might be a more programmatic way to do this. Potentially mapping the game objects from the database into nested objects for each player in the game.... that would, in theory, allow us to just loop through each player's object from the game and apply them to the correct player.

/**
  * applyGame - used in buildTable to apply the game to the table
  * @param {object} table - the current table
  * @param {object} game - the current game to be applied
*/
export const applyGame = (table, game) => {

  let { p1, p2, p1Score, p2Score, p1Shots, p2Shots, p1Poss, p2Poss, p1OnGoal, p2OnGoal, p1Reds, p1Yellows, p2Reds, p2Yellows } = game;


  let player1 = table[p1];
  let player2 = table[p2];

  if (p1Score && p2Score) {

    p1Score === p2Score ? (
      player1.draws++,
      player2.draws++
    ) : p1Score > p2Score ? (
      player1.wins++,
      player2.losses++
    ) : (
      player2.wins++,
      player1.losses++
    );
    player1.goalsFor += p1Score;
    player2.goalsFor += p2Score;
    player1.goalsAgainst += p2Score;
    player2.goalsAgainst += p1Score;
    player1.shots += p1Shots;
    player2.shots += p2Shots;
    player1.onGoal += p1OnGoal;
    player2.onGoal += p2OnGoal;
    player1.reds += p1Reds;
    player2.reds += p2Reds;
    player1.yellows += p1Yellows;
    player2.yellows += p2Yellows;
    player1.poss = player1.poss === null ? p1Poss : (player1.poss + p1Poss) / (player1.wins + player1.losses + player1.draws);
    player2.poss = player2.poss === null ? p2Poss : (player1.poss + p1Poss) / (player1.wins + player1.losses + player1.draws);
    player1.points = (player1.wins * 3 + player1.draws);
    player2.points = (player2.wins * 3 + player2.draws);
    player1.goalDiff = (player1.goalsFor - player1.goalsAgainst);
    player2.goalDiff = (player2.goalsFor - player2.goalsAgainst);
  }
  return table;
};


// NOTE: Currently not being used:
  // Need to figure out how to get the react-bootstrap - tooltips to work.
  // Seemingly I need to build a component wrapped in a <th>... but the first attempt didn't work.
export const mapHeaders = (header) => {
  switch (header) {
  case 'P':
    return 'Points';
  case 'W':
    return 'Wins';
  case 'L':
    return 'Losses';
  case 'D':
    return 'Draws';
  case 'GF':
    return 'Goals For';
  case 'GA':
    return 'Goals Against';
  case 'GD':
    return 'Goal Differencial';
  case 'S':
    return 'Shots';
  case 'OT':
    return 'Shots On Target';
  case 'Y':
    return 'Yellows';
  case 'R':
    return 'Reds';
  case 'Po':
    return 'Possession';
  default:
    return 'Player';
  }
};
