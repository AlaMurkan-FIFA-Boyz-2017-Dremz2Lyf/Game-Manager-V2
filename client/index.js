// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';

// css
import '../styles/styles.css';

// Bring in our router
import Routes from './router';

// Grab our root reducer and our store from their files
import rootReducer from './reducers/index.js';
import { store } from './store';

// Found this... Needed for hot module reloading.
if (module.hot) {
  module.hot.accept();
}

ReactDOM.render(
  <Provider store={store}>
    <Routes />
  </Provider>,
  document.getElementById('app')
);
