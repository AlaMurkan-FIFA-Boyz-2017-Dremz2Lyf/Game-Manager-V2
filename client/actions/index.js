// dependancies
import axios from 'axios';

// Action Types
import { SET_ERRORED, SET_LOADING, RECEIVE } from './types';

// Utilities
import { normalize } from '../utilities';

export const setErrored = (stateKey, data) => {
  // NOTE: Still not 100% sure if this is going to work for all errors.... but it might! :D
  let error = data.status ? {status: data.status, message:Â data.data} : {message: data.message, source: data.stack};

  return {
    type: SET_ERRORED,
    payload: {
      data: error,
      stateKey
    }
  };
};


export const setLoading = (stateKey, data) => {
  return {
    type: SET_LOADING,
    payload: {
      data,
      stateKey
    }
  };
};

export const receive = (stateKey, data) => {
  // receive is set up to accept an array or an object.
    // If it is an array we do the normalization here.
  if (Array.isArray(data)) {
    data = normalize(data);
  }

  // and send off our action with the proper payload
  return {
    type: RECEIVE,
    payload: {
      data,
      stateKey
    }
  };
};

export const fetch = (stateKey, data = {}) => {
  let request = axios.get(`/${stateKey}`, {params: data});

  return dispatch => {
    dispatch(setLoading(stateKey, true));

    return request.then(response => response.data).then(data => {
      dispatch(receive(stateKey, data));
    }).catch(error => {
      // This should handle any response errors, but more than likely not anything else.... D:
      let data = error.response ? error.response : error;
      dispatch(setErrored(stateKey, data));
    }).then(dis => {
      dispatch(setLoading(stateKey, false));
    });
  };
};

// In an effort to minimize request traffic, the server will respond with the updated or created data object we are currently working with. Instead of fetching all of that corresponding data we need after the post. We will just pass the updated data through to the receive action, by using getState from redux-thunk.

export const create = (stateKey, data) => {
  let post = axios.post(`/${stateKey}`, data);

  return (dispatch, getState) => {
    dispatch(setLoading(stateKey, true));

    return post.then(response => normalize(response.data)).then(updated => {

      // Grab the specific subState for the stateKey we are working with during this action
      let subState = getState().data[stateKey];

      // Take the subState and the updated data and create a new object
      let updatedSubState = {...subState, ...updated};

      // Pass that updated subState into the receive action
      return dispatch(receive(stateKey, updatedSubState));
    }).catch(error => {

      // If there was an error in the put, we grab the response from the server and pass it along to our setErrored action creator.
      let data = error.response;
      return dispatch(setErrored(stateKey, data));
    }).then(dis =>

      // Once everything is done, we set our loading flag to false.
      dispatch(setLoading(stateKey, false))
    );
  };
};

export const update = (stateKey, data) => {
  let put = axios.put(`/${stateKey}`, data);

  return (dispatch, getState) => {
    // Set our loading flag for the specific stateKey we are working with
    dispatch(setLoading(stateKey, true));

    // Then handle the async request.
    return put.then(response => normalize(response.data)).then(updated => {

      // Grab the specific subState for the stateKey we are working with during this action
      let subState = getState().data[stateKey];

      // Take the subState and the updated data and create a new object
      let updatedSubState = {...subState, ...updated};

      // Pass that updated subState into the receive action
      return dispatch(receive(stateKey, updatedSubState));
    }).catch(error => {

      // If there was an error in the put, we grab the response from the server and pass it along to our setErrored action creator.

      let data = error.response;
      return dispatch(setErrored(stateKey, data));
    }).then(dis =>

      // Once everything is done, we set our loading flag to false.
      dispatch(setLoading(stateKey, false))
    );
  };
};
