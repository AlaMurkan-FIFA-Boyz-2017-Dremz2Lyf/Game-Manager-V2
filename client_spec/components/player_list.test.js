import React from 'react';
import { Provider } from 'react-redux';

import PlayerList from '../../client/components/player_list';
import { store } from '../../client/store';


describe('PlayerList', () => {
  let mockMove = jest.fn();
  let wrapper = shallow( <PlayerList searchValue={''} all={false} move={mockMove}/> );

  test('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should not render the CreatePlayer component if all is false', () => {

    expect(wrapper.find('CreatePlayer').length).toBe(0);
  });

  test('should render Components correctly based on props', () => {

    expect(wrapper.find('CreatePlayer').length).toBe(0);
    expect(wrapper.find('PlayerListItem').length).toBe(0);

    wrapper.setProps({
      all: true,
      players: mockData.players
    });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Connect(CreatePlayer)').length).toBe(1);
    expect(wrapper.find('PlayerListItem').length).toBe(3);

  });


});
