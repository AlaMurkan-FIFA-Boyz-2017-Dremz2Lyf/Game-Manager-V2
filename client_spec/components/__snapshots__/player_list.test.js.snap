exports[`PlayerList should match the snapshot 1`] = `
<ListGroup
  bsClass="list-group" />
`;

exports[`PlayerList should render Components correctly based on props 1`] = `
<ListGroup
  bsClass="list-group">
  <PlayerListItem
    move={[Function]}
    player={
      Object {
        "createdAt": "2016-12-25T19:31:48",
        "draws": 0,
        "goalsAgainst": 0,
        "goalsFor": 0,
        "id": 1,
        "isTeam": 0,
        "losses": 0,
        "onGoal": 0,
        "possession": 0,
        "reds": 0,
        "shots": 0,
        "updatedAt": null,
        "username": "Alice",
        "wins": 0,
        "yellows": 0,
      }
    } />
  <PlayerListItem
    move={[Function]}
    player={
      Object {
        "createdAt": "2016-12-25T19:31:48",
        "draws": 0,
        "goalsAgainst": 0,
        "goalsFor": 0,
        "id": 2,
        "isTeam": 0,
        "losses": 0,
        "onGoal": 0,
        "possession": 0,
        "reds": 0,
        "shots": 0,
        "updatedAt": null,
        "username": "Gilbert",
        "wins": 0,
        "yellows": 0,
      }
    } />
  <PlayerListItem
    move={[Function]}
    player={
      Object {
        "createdAt": "2016-12-25T19:31:48",
        "draws": 0,
        "goalsAgainst": 0,
        "goalsFor": 0,
        "id": 3,
        "isTeam": 0,
        "losses": 0,
        "onGoal": 0,
        "possession": 0,
        "reds": 0,
        "shots": 0,
        "updatedAt": null,
        "username": "Bob",
        "wins": 0,
        "yellows": 0,
      }
    } />
  <Connect(CreatePlayer) />
</ListGroup>
`;
